File: src/161. One Edit Distance/161.cpp
/*
 * File: 161.cpp
 * Created Date: 2021-01-23
 * Author: Nick Cao(caoyxsh@outlook.com)
 * Brief: https://leetcode.com/problems/one-edit-distance/
 * -----
 * Last Modified: 2021-01-23T08:20:21-08:00
 * Modified By: Nick Cao(caoyxsh@outlook.com)
 */

/**
 * @brief My approach, use two pointer, check from being and end of two strings
 * Time: O(n)
 * Space: O(n)
 */
class Solution {
    bool isOneEditDistance(string& s, string& t, int sl, int sr, int tl, int tr) {
        if(s.empty()&&t.empty()) return false;
        if(s.empty()) return t.size() < 2;
        if(t.empty()) return s.size() < 2;
        while(sl < sr && tl < tr && s[sl] == t[tl]){
            ++sl; ++tl;
            // printf("l\n");
        }
        while(sl < sr && tl < tr && s[sr-1] == t[tr-1]){
            --sr; --tr;
            // printf("r\n");
        }
        int s_len = sr-sl;
        int t_len = tr-tl;
        // printf("%d %d\n",s_len,t_len);
        if(s_len == t_len) return s_len == 1;
        else return abs(s_len-t_len) == 1 && (s_len == 0 || t_len == 0);
    }
public:
    bool isOneEditDistance(string s, string t) {
        return isOneEditDistance(s,t,0,s.size(),0,t.size());
    }
};

File: src/161.OneEditDistance/161.java
/*
 * File: 161.java
 * Created Date: 2021-01-23
 * Author: Nick Cao(caoyxsh@outlook.com)
 * Brief: https://leetcode.com/problems/one-edit-distance/
 * -----
 * Last Modified: 2021-01-23T08:21:55-08:00
 * Modified By: Nick Cao(caoyxsh@outlook.com)
 */

/**
 * @brief 
 * Time: O(n)
 * Space: O(n)
 */
class Solution {
    public boolean isOneEditDistance(String s, String t) {
      int ns = s.length();
      int nt = t.length();
  
      // Ensure that s is shorter than t.
      if (ns > nt)
        return isOneEditDistance(t, s);
  
      // The strings are NOT one edit away distance  
      // if the length diff is more than 1.
      if (nt - ns > 1)
        return false;
  
      for (int i = 0; i < ns; i++)
        if (s.charAt(i) != t.charAt(i))
          // if strings have the same length
          if (ns == nt)
            return s.substring(i + 1).equals(t.substring(i + 1));
          // if strings have different lengths
          else
            return s.substring(i).equals(t.substring(i + 1));
  
      // If there is no diffs on ns distance
      // the strings are one edit away only if
      // t has one more character. 
      return (ns + 1 == nt);
    }
  }